### PR TITLE
installer ux: add K8s client progress into installer status screen

### LINKED ISSUE #561: Add Kubernetes client into progress screen to track pod/component statuses
Give more rich feedback about when containers are downloading and launching.

### PR DESCRIPTION
**Pull request by LyonesGamer**
Fixes #561.

 * Separates the Tectonic and Terraform status APIs so they don't rely on each other
 * Splits the "Starting Tectonic Console" status check into multiple Kubernetes and Tectonic components

Still in progress.

### CODE REVIEW

**Comment by robszumski** [comment]
Once you get to the more detailed installer screens, would you mind adding a few screenshots here so we can track progress?

**Comment by LyonesGamer** [comment]
Currently blocked on #1229.

**Comment by LyonesGamer** [comment]
@robszumski 

Here's what I've got so far:

screenshot from 2017-06-29 14-21-31

**Comment by tlwu2013** [comment]
Update the mocks here. Please just focus on the component updates' styling for this PR and ignore all other differences (e.g. "progress bar", "download assets button", etc).

I put some notes on the mocks, please let me know if any questions, thanks!

component statuses start installation__step-1--redline

component statuses start installation__step-2--redline

component statuses start installation__done--redline

**Comment by LyonesGamer** [comment]
Here's what I've done so far:

Still starting...
screenshot from 2017-06-30 14-35-00

Some of them running...
screenshot from 2017-06-30 14-40-26

Finished!
screenshot from 2017-06-30 14-43-52

I didn't see the bit about hiding them when they're all done, I'll get to work on that now.

**Comment by LyonesGamer** [comment]
ok to test

**Comment by LyonesGamer** [comment]
ok to test

**Comment by LyonesGamer** [comment]
ok to test

**Comment by LyonesGamer** [comment]
I believe this PR to be ready for a full review, although it is still blocked on #1229.

**Comment by LyonesGamer** [comment]
ready to test

File: installer/api/handlers_tectonic.go [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+	}
+	etcd, ok := val.(bool)
+	if !ok {
+		return false, errors.New("Could not parse experimental flag as bool")
+	}
+	return etcd, nil
+}
+
+func newK8sClient(ex terraform.Executor) (*kubernetes.Clientset, error) {
+	kCfgPath := filepath.Join(ex.WorkingDirectory(), "generated/auth/kubeconfig")
```
Comment (squat) [inline installer/api/handlers_tectonic.go]: Can you make this "generated/auth/kubeconfig" a const with a comment

File: installer/api/handlers_tectonic.go [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+	"strings"
+
+	"errors"
+
+	"github.com/coreos/tectonic-installer/installer/pkg/tectonic"
+	"github.com/coreos/tectonic-installer/installer/pkg/terraform"
+	"github.com/dghubble/sessions"
+)
+
+// Status represents the individual state of a single service
```
Comment (squat) [inline installer/api/handlers_tectonic.go]: Please clarify are these systemd services, k8s or both?

File: installer/api/handlers_tectonic.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+	Kubernetes Status                 `json:"kubernetes"`
+	HasEtcd    bool                   `json:"hasetcd"`
+	Etcd       Status                 `json:"etcd"`
+	Identity   Status                 `json:"identity"`
+	Ingress    Status                 `json:"ingress"`
+	Console    tectonic.ServiceStatus `json:"console"`
+	Tectonic   Status                 `json:"tectonic"` // All other Tectonic services
+}
+
+func isEtcdHosted(ex terraform.Executor) (bool, error) {
```
Comment (squat) [inline installer/api/handlers_tectonic.go]: Please add comments to all these functions. Also, etcd is always hosted, the question is whether it is "self" hosted.

File: installer/api/handlers_tectonic.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+	etcd, ok := val.(bool)
+	if !ok {
+		return false, errors.New("Could not parse experimental flag as bool")
+	}
+	return etcd, nil
+}
+
+func newK8sClient(ex terraform.Executor) (*kubernetes.Clientset, error) {
+	kCfgPath := filepath.Join(ex.WorkingDirectory(), "generated/auth/kubeconfig")
+
```
Comment (squat) [inline installer/api/handlers_tectonic.go]: Could you stat this filepath to make sure the kubeconfig exists already and return an error if it does not. This would give you information to branch off of in tectonicStatus.

File: installer/api/handlers_tectonic.go [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+		Ingress:    Status{Success: false, Failed: false},
+		Console:    tectonic.ConsoleHealth(nil, input.TectonicDomain),
+		Tectonic:   Status{Success: false, Failed: false},
+	}
+
+	client, err := newK8sClient(*ex)
+	if err != nil {
+		// This is probably because the kubeconfig hasn't been generated yet; assume services haven't started
+		// TODO: Better error handling to make sure it's not a different issue
+		return services, nil
```
Comment (squat) [inline installer/api/handlers_tectonic.go]: Rather than return nil, return a known error and branch off of that in the function that calls this func. Otherwise you're flying blind here

File: installer/api/handlers_tectonic.go [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+	}
+
+	pods, err := client.CoreV1().Pods("").List(v1.ListOptions{})
+	if err != nil {
+		// APIServer probably hasn't started yet; assume services haven't started
+		// TODO: Better error handling to make sure it's not a different issue
+		return services, nil
```
Comment (squat) [inline installer/api/handlers_tectonic.go]: Same as above

File: installer/api/handlers_tectonic.go [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+	}
+
+	etcd, err := isEtcdHosted(*ex)
+	if err != nil {
+		return services, newInternalServerError("Error reading TF Vars: %s", err)
```
Comment (squat) [inline installer/api/handlers_tectonic.go]: This error should communicate that the function "failed to determine if etcd is self hosted".

File: installer/api/handlers_tectonic.go [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+// Status represents the individual state of a single service
+type Status struct {
+	Success bool `json:"success"`
+	Failed  bool `json:"failed"`
+}
+
+// Services represents whether several Tectonic components have yet booted (or failed)
+type Services struct {
+	Kubernetes Status                 `json:"kubernetes"`
+	HasEtcd    bool                   `json:"hasetcd"`
```
Comment (squat) [inline installer/api/handlers_tectonic.go]: Better name for this member.

File: installer/api/handlers_tectonic.go [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+	services.HasEtcd = etcd
+
+	services.Kubernetes.Success = true
+	services.Tectonic.Success = true
+
+	var kubeServs []v1.Pod
+	var tectServs []v1.Pod
+
+	for _, p := range pods.Items {
+		name := p.ObjectMeta.Name
```
Comment (squat) [inline installer/api/handlers_tectonic.go]: I don't think this is a good heuristic. Instead switch on the namespace. "kube-system" or "tectonci-system" and no default

File: installer/api/handlers_tectonic.go [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+	if err != nil {
+		return services, newInternalServerError("Error reading TF Vars: %s", err)
+	}
+
+	services.HasEtcd = etcd
+
+	services.Kubernetes.Success = true
+	services.Tectonic.Success = true
+
+	var kubeServs []v1.Pod
```
Comment (squat) [inline installer/api/handlers_tectonic.go]: It is confusing to call these things services in the context of k8s since a service is a particular concept. just call them pods.

File: installer/api/handlers_tectonic.go [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+type TerraformStatus struct {
+	Status string `json:"status"`
+	Output string `json:"output,omitempty"`
+	Error  string `json:"error,omitempty"`
+	Action string `json:"action"`
+}
+
+func terraformStatus(session *sessions.Session, ex *terraform.Executor, exID int) (TerraformStatus, error) {
```
Comment (squat) [inline installer/api/handlers_tectonic.go]: Document this function please

File: installer/api/handlers_tectonic.go [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+	// Retrieve Terraform's status and output.
+	status, err := ex.Status(exID)
+	if status == terraform.ExecutionStatusUnknown {
+		return TerraformStatus{}, newBadRequestError("Could not retrieve TerraForm execution's status: %s", err)
+	}
+	output, _ := ex.Output(exID)
+	outputBytes, _ := ioutil.ReadAll(output)
```
Comment (squat) [inline installer/api/handlers_tectonic.go]: Why are we ignoring these errors?

File: installer/frontend/components/tf-poweron.jsx [hunk 13]
```diff
@@ -81,27 +83,73 @@ class TF_PowerOn extends React.Component {
...
+
+      const components = [
+        {key: 'kubernetes', name: 'Kubernetes'},
+        {key: 'identity', name: 'Tectonic Identity'},
+        {key: 'ingress', name: 'Tectonic Ingress Controller'},
+        {key: 'console', name: 'Tectonic Console'},
+        {key: 'tectonic', name: 'other Tectonic services'},
+      ];
+
+      if (tectonic.hasetcd) {
```
Comment (squat) [inline installer/frontend/components/tf-poweron.jsx]: Please rename hadetcd to something like etcdIsSeñfHosted.

File: installer/api/handlers_tectonic.go [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+	Console    tectonic.ServiceStatus `json:"console"`
+	Tectonic   Status                 `json:"tectonic"` // All other Tectonic services
+}
+
+func isEtcdHosted(ex terraform.Executor) (bool, error) {
+	data, err := ex.LoadVars()
+	if err != nil {
+		return false, err
+	}
+	val, ok := data["tectonic_experimental"]
```
Comment (kans) [inline installer/api/handlers_tectonic.go]: Using "tectonic_experimental" for determining if etcd is self hosted is brittle. - the meaning of this flag has changed every release.

Comment (LyonesGamer) [inline installer/api/handlers_tectonic.go]: Is there a better way to do it? This is how both Quentin and Geoff said to find out.

File: installer/assets/frontend/css/styles.css [hunk 15]
```diff
@@ -156,6 +156,28 @@ input[type="text"].wiz-node-field {
...
+  font-size: 14px;
+}
+
+.service-launch-progress__step {
+  list-style: none;
+  margin: 0 0 8px 15px;
+  padding: 0;
+}
+
+.service-launch-progress__step:first-child {
```
Comment (kans) [inline installer/assets/frontend/css/styles.css]: We generally try to avoid CSS pseudo-classes because they cause so much trouble.

File: installer/api/handlers_tectonic.go (hunk 14 above)
Comment (kans) [inline installer/api/handlers_tectonic.go]: The "right" way to check would be to see if we are running etcd on k8s.  I'm not sure its worth effort honestly.

File: installer/api/handlers_tectonic.go (hunk 12 above)
Comment (LyonesGamer) [inline installer/api/handlers_tectonic.go]: Good question, I'm not sure; this section was just taken from the previous terraformStatusHandler() function.

Comment (squat) [inline installer/api/handlers_tectonic.go]: ok, let's handle these errors then. We can afford to be careful and verbose here.

File: installer/api/handlers_tectonic.go [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+// Services represents whether several Tectonic components have yet booted (or failed)
+type Services struct {
+	Kubernetes Status                 `json:"kubernetes"`
+	HasEtcd    bool                   `json:"hasetcd"`
+	Etcd       Status                 `json:"etcd"`
+	Identity   Status                 `json:"identity"`
+	Ingress    Status                 `json:"ingress"`
+	Console    tectonic.ServiceStatus `json:"console"`
```
Comment (kans) [inline installer/api/handlers_tectonic.go]: `Console` is a different type than the rest of these fields.  Any chance we could clean this up while we are here?

File: installer/frontend/components/tf-poweron.jsx [hunk 17]
```diff
@@ -81,27 +83,73 @@ class TF_PowerOn extends React.Component {
...
+
       consoleSubsteps.push(
-        <WaitingLi done={dnsReady} cancel={tfError} key="dns">
+        <WaitingLi done={dnsReady} key='dnsReady'>
           <span title={msg}>{msg}</span>
         </WaitingLi>
       );
-      msg = `Starting Tectonic console`;
+
+      const components = [
```
Comment (kans) [inline installer/frontend/components/tf-poweron.jsx]: I would expect components in this context to refer to React Components.

File: installer/frontend/components/tf-poweron.jsx [hunk 18]
```diff
@@ -81,27 +83,73 @@ class TF_PowerOn extends React.Component {
...
+        {key: 'tectonic', name: 'other Tectonic services'},
+      ];
+
+      if (tectonic.hasetcd) {
+        components.splice(1, 0, {key: 'etcd', name: 'Etcd'});
+      }
+
+      let allDone = true, anyFailed = false;
+
+      for (let c of components) {
```
Comment (kans) [inline installer/frontend/components/tf-poweron.jsx]: Its hard to follow this logic.  If am I reading this correctly, `anyFailed` means any one component has a failed status and `allDone` means every component has a success status besides `tectonic` which should have the ready status.

Since we don't care about efficiency at all, you could write `anyFailed` as eg:

`const anyFailed = _.some(components, c => tectonic[c.key].failed)`

... and the equivalent for `allDone`.  Cleaning up the actual response from the back end would also make life easier here.

File: installer/api/handlers_tectonic.go [hunk 19]
```diff
@@ -0,0 +1,240 @@
+package api
```
Comment (squat) [inline installer/api/handlers_tectonic.go]: I just realized that the formatting for this whole file is a little off. please format using `gofmt -s -w`

File: installer/assets/frontend/css/styles.css (hunk 15 above)
Comment (LyonesGamer) [inline installer/assets/frontend/css/styles.css]: How would you recommend I rewrite that section? As it is, it basically just mirrors the .wiz-launch-progress classes.

File: installer/api/handlers_tectonic.go (hunk 19 above)
Comment (LyonesGamer) [inline installer/api/handlers_tectonic.go]: Gofmt didn't cause any changes.

I have my IDE set so that it should automatically run gofmt anyway.

File: installer/api/handlers_tectonic.go (hunk 16 above)
Comment (LyonesGamer) [inline installer/api/handlers_tectonic.go]: The primary issue is that the DNS waiting icon depends on the ServiceStatus "instance" field. How would you recommend separating that?

File: installer/api/handlers_tectonic.go (hunk 19 above)
Comment (squat) [inline installer/api/handlers_tectonic.go]: ok good. Looks like github diff view just renders tabs oddly.

File: installer/assets/frontend/css/styles.css (hunk 15 above)
Comment (kans) [inline installer/assets/frontend/css/styles.css]: Keep track of the position in JS when you render the the Components and add classes to the fist/last ones.  You can punt on it if its too much work.

**Comment by robszumski** [comment]
How is this progressing?

**Comment by LyonesGamer** [comment]
@robszumski It should get another round of review, but it looks like I fixed most of the issues with it. It's still blocked on #1229, though. Without that PR, the installer will not build.

**Comment by sym3tri** [comment]
@squat can you take this over since @LyonesGamer will be out for a few days. Hopefeully we can get it in before Friday unless there are more pressing things.

**Comment by squat** [comment]
@sym3tri gladly taking over. Should be good to merge by early tomorrow.

**Comment by ggreer** [comment]
There's not enough time today to deal with this. Matt & I will mess with it on Monday.

**Comment by squat** [comment]
after a few hours of work, this PR still needs more work. Much of the logic is broken and it needs more thorough testing of both frontend and backend pieces. Deferring until Monday.

**Comment by ggreer** [comment]
This works OK now, but it has one major issue: We don't know how many pods there will be, so statuses can flap between success and pending.

Fixes could include:

- Bubbling up all pods & status to the front end, and showing more detailed info there (eg: "5/8 ingress pods running", and both 5 and 8 increase).
- Pinning statuses so that as soon as we see one success, we never return that status to pending.
- Waiting for n successes in a row before showing a status as succeeded.

I'm not sure what the best thing to do is, but I think the current UI is better than what's in master.

**Comment by ggreer** [comment]
Screenshots:
screenshot from 2017-07-25 18-38-43
screenshot from 2017-07-25 18-38-49
screenshot from 2017-07-25 18-42-29
screenshot from 2017-07-25 18-44-20
screenshot from 2017-07-25 18-47-24

**Comment by ggreer** [comment]
Now that I made the styling closer to the mocks, flapping isn't quite as obvious:
screenshot from 2017-07-26 10-46-19

**Comment by robszumski** [comment]
> Now that I made the styling closer to the mocks, flapping isn't quite as obvious:

Thank you, this was a conscious decision during the design process for this reason :) Hat tip to @tlwu2013
